### PR TITLE
query func params

### PR DESCRIPTION
**Pull request by francescov1**
- Adds support for proper query function parameter typing,
- Refactor of generated method and static type structure (same as query functions now). This removes the need to use "fake this" param on methods & statics, just need to cast your Schema.methods or Schema.statics to the new Methods and Statics interfaces when initializing.

@HoldYourWaffle Wanted to merge in before end of weekend, if you have any feedback lets start a new PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,36 +137,33 @@ const UserSchema = new Schema({
...
 
-// method functions
+// method functions, cast to UserStatics
 UserSchema.methods = {
-  isMetadataString(this: UserDocument) {
+  isMetadataString() {
     return typeof this.metadata === "string";
   }
-};
+} as UserMethods;
```
Comment (HoldYourWaffle) [inline README.md:151]: Why was this changed to a cast?
I don't see any advantages (but that could be me), and it reduces type-safety.

File: README.md [hunk 2]
```diff
@@ -137,36 +137,33 @@ const UserSchema = new Schema({
...
-// static functions
+// static functions, cast to UserStatics
 UserSchema.statics = {
   // friendUids could also use the type `ObjectId[]` here
-  async getFriends(this: UserModel, friendUids: UserDocument["_id"][]) {
+  async getFriends(friendUids: UserDocument["_id"][]) {
     return await this.aggregate([{ $match: { _id: { $in: friendUids } } }]);
   }
-};
+} as UserStatics;
```
Comment (HoldYourWaffle) [inline README.md:159]: Same as above, why a cast?

File: README.md [hunk 3]
```diff
@@ -137,36 +137,33 @@ const UserSchema = new Schema({
...
-const queryFuncs: UserQueries = {
+// query functions, cast to UserQueries
+UserSchema.query = {
   populateFriends() {
     return this.populate("friends.uid", "firstName lastName");
   }
-};
-
-UserSchema.query = queryFuncs;
+} as UserQueries;
```
Comment (HoldYourWaffle) [inline README.md:166]: Same as above, why a cast?

File: README.md [hunk 4]
```diff
@@ -223,13 +223,13 @@ export type UserFriendDocument = mongoose.Types.Subdocument & {
...
-export type UserDocument = mongoose.Document & {
-  metadata?: any;
-  friends: mongoose.Types.DocumentArray<UserFriendDocument>;
-  city: {};
-  name: any;
-  isMetadataString: (this: any) => boolean;
-} & User;
+export type UserDocument = mongoose.Document &
+  UserMethods & {
+    metadata?: any;
```
Comment (HoldYourWaffle) [inline README.md:228]: Is this a "user can put anything in it" kind of `any` type?

Might be worth looking into [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) in that case, so users can provide their type.

File: README.md [hunk 5]
```diff
@@ -223,13 +223,13 @@ export type UserFriendDocument = mongoose.Types.Subdocument & {
...
+    friends: mongoose.Types.DocumentArray<UserFriendDocument>;
+    city: {};
+    name: any;
```
Comment (HoldYourWaffle) [inline README.md:231]: Why is this `any`?

File: README.md [hunk 6]
```diff
@@ -280,6 +280,5 @@ async function editEmail(user: UserDocument, newEmail: string): UserDocument {
...
-- [ ] Query function parameters are typed using a rest parameter `(...args: any[])`, this needs to be fine tunned to use the actual parameters and types.
 - [ ] Top-level schema fields that refer to the schema itself (i.e. an array of User friend IDs at the root of the User schema) will be typed as the barebones Mongoose-less Schema interface, rather than the Document type (in example above, would refer to `User` instead of `UserDocument`). This is because the Document type is a TS type, rather than an interface, thus it cannot reference itself. This edge-case only really arises to users if they populate that specific property, otherwise this would references an ObjectId in both cases.
-- [ ] Generating types for specifically the method and static functions objects. This is how query functions are currently handled, it removes the need to fill in the "fake this" parameter for each function.
+- [ ] Cut down node_modules by using users own local installation (i.e. mongoose) and stripping oclif if possible.
```
Comment (HoldYourWaffle) [inline README.md:284]: Wouldn't the local installation part just be a peer dependency?

File: src/helpers/tests/artifacts/user.ts [hunk 7]
```diff
@@ -36,36 +36,33 @@ const UserSchema = new Schema({
...
 });
 
 // method functions
 UserSchema.methods = {
-  isMetadataString(this: UserDocument) {
+  isMetadataString() {
     return typeof this.metadata === "string";
   }
-};
+} as UserMethods;
```
Comment (HoldYourWaffle) [inline src/helpers/tests/artifacts/user.ts:50]: Same as earlier, why a cast?

File: src/helpers/tests/artifacts/user.ts [hunk 8]
```diff
@@ -36,36 +36,33 @@ const UserSchema = new Schema({
...
 
 // static functions
 UserSchema.statics = {
   // friendUids could also use the type `ObjectId[]` here
-  async getFriends(this: UserModel, friendUids: UserDocument["_id"][]) {
+  async getFriends(friendUids: UserDocument["_id"][]) {
     return await this.aggregate([{ $match: { _id: { $in: friendUids } } }]);
   }
-};
+} as UserStatics
```
Comment (HoldYourWaffle) [inline src/helpers/tests/artifacts/user.ts:58]: Same as above, why a cast?

File: src/helpers/tests/artifacts/user.ts [hunk 9]
```diff
@@ -36,36 +36,33 @@ const UserSchema = new Schema({
...
-const queryFuncs: UserQueries = {
+// query functions
+UserSchema.query = {
   populateFriends() {
     return this.populate("friends.uid", "firstName lastName");
   }
-};
-
-UserSchema.query = queryFuncs;
+} as UserQueries
```
Comment (HoldYourWaffle) [inline src/helpers/tests/artifacts/user.ts:65]: Same as above, why a cast?

**Review by HoldYourWaffle — COMMENTED** [review]
No major comments here 🙃 

I'm not entirely 100% exactly sure what all these changes are (too much of a mongoose noob), but it looks like the generated output has changed quite a bit. Does that mean that this is a breaking change?

File: README.md (hunk 1 above)
Comment (francescov1) [reply README.md:151]: Casting enforces the function types in this case. If you take a look at the mongoose typings for these schema properties (methods, statics, and query) they look like this:

```
/** Object of currently defined methods on this schema. */
    methods: {
      [F in keyof T]: T[F]
    };
    /** Object of currently defined statics on this schema. */
    statics: any;
    /** Object of currently defined query helpers on this schema. */
    query: any;
```

This casting doesn't reduce the type-safety, if anything it improves it. It can be seen as more of "type assertion" then typecasting. It also substantially reduces the changes a user needs to make to use the library (no need for `this: UserDocument`)

File: README.md (hunk 4 above)
Comment (francescov1) [reply README.md:228]: The `any` comes from `metadata` being defined as a `Mixed` mongoose type.

File: README.md (hunk 5 above)
Comment (francescov1) [reply README.md:231]: For `metadata` see https://github.com/Bounced-Inc/mongoose-tsgen/pull/18#discussion_r542964445. `name: any` is because `name` is declared as a virtual property. These don't have defined types so cannot be pulled from the Mongoose schemas. We could try to grab these with ts-morph in the future.

File: README.md (hunk 6 above)
Comment (francescov1) [reply README.md:284]: Yes 100%, it was late and I was blanking on the term 😂

**Comment by francescov1** [comment]
Nope, shouldnt be breaking (essentially just splits out static and method functions to use their own interfaces, rather than embedding them into the model and document interfaces respectively). These new standalone interfaces can then be used to enforce a stronger typing on the `Schema.methods` and `Schema.statics` objects and remove the need to provide the "fake this".

File: README.md (hunk 6 above)
Comment (francescov1) [reply README.md:284]: @HoldYourWaffle any thoughts on which dependencies can be set as peer? Don't have a ton of experience with this, but I've noticed that the node_modules is huge and a good chunk of the dependencies used are most often going to be present in users own projects (mongoose for sure, eslint, prettier, ts-node, etc.)
